src/generic/oomph_utilities.cc

```diff
@@ -1306,6 +1306,50 @@ namespace oomph
         return double(t) / double(CLOCKS_PER_SEC);
       }
     }
+
+    /// Returns a nicely formatted string from an input time in seconds;
+    /// the format depends on the size of time, e.g.:
+    /// 86510 will be printed as 1d 1m:50
+    ///  3710 will be printed as 1h:01:50
+    ///   700 will be printed as 11m:40
+    ///    59 will be printed as 59s
+    std::string convert_secs_to_formatted_string(
+      const long unsigned& time_in_sec)
+    {
+      std::ostringstream ss;
+
+      unsigned sec_within_day = time_in_sec % (3600 * 24);
+
+      unsigned time_days = time_in_sec / (3600 * 24);
+      unsigned time_hrs = sec_within_day / 3600;
+      unsigned time_min = (sec_within_day % 3600) / 60;
+      unsigned time_sec = (sec_within_day % 3600) % 60;
+
+      if (time_days > 0)
+      {
+        ss << time_days << "d ";
+      }
+
+      if (time_hrs > 0)
+      {
+        ss << time_hrs << "h:";
+        ss << std::setw(2) << std::setfill('0');
+        ss << time_min << ":";
+        ss << time_sec;
+      }
+      else if (time_min > 0)
+      {
+        ss << time_min << "m:";
+        ss << std::setw(2) << std::setfill('0');
+        ss << time_sec;
+      }
+      else
+      {
+        ss << time_sec << "s";
+      }
+
+      return ss.str();
+    }
```

Comment:
This can probably be simplified using the built-in `sprintf` function.

```cpp
// Type alias for unsigned long long (will locally override the "time_t" in <ctime>)
using time_t = unsigned long long;    

// -------------------------------------- TODO: Delete
#include <climits>
time_t time_in_sec = ULLONG_MAX;
// -------------------------------------- TODO: Delete

// Calculate quantities
time_t days = time_in_sec / (3600 * 24);
time_t hours = (time_in_sec % (3600 * 24)) / 3600;
time_t minutes = (time_in_sec % 3600) / 60;
time_t seconds = time_in_sec % 60;
    
// Comment out this #define to see the output of the other block
#define WHITESPACE_DELIMETER
#ifdef WHITESPACE_DELIMETER

// This version is clear due to the visual whitespace break between quantities.
// Example: ULLONG_MAX -> "213503982334601h 07h 00m 15s".
char time[100];
sprintf(time, "%lluh %02lluh %02llum %02llus", days, hours, minutes, seconds);
std::string time_as_string(time);

#else

// Slightly harder to read due to the lack of a visual delimeter in the second 
// set of quantities.
// Example: ULLONG_MAX -> "213503982334601h 07h:00m:15s".
char time[100];
sprintf(time, "%lluh %02lluh:%02llum:%02llus", days, hours, minutes, seconds);
std::string time_as_string(time);

#endif

// -------------------------------------- TODO: Delete
std::cout << time_as_string << std::endl;
// -------------------------------------- TODO: Delete
```

Reply:
The above looks slightly messy because of the comparison but the finished product would look like the following

```cpp
std::string convert_secs_to_formatted_string(const unsigned long long& time_in_sec) {
    // Type alias for unsigned long long (will locally override the "time_t" in <ctime>)
    using time_t = unsigned long long;    
    
    // Calculate quantities
    time_t days = time_in_sec / (3600 * 24);
    time_t hours = (time_in_sec % (3600 * 24)) / 3600;
    time_t minutes = (time_in_sec % 3600) / 60;
    time_t seconds = time_in_sec % 60;
            
    // This version is clear due to the visual whitespace break between quantities.
    // Example: ULLONG_MAX -> "213503982334601h 07h 00m 15s".
    char time[100];
    sprintf(time, "%lluh %02lluh %02llum %02llus", days, hours, minutes, seconds);
    return std::string{time};
}
```

Reply:
I have to disagree - with this implementation, convert_secs_to_formatted_string(1) would print:
0d 00h 00m 01s
which is unnecessarily hard to read, whereas my implementation would print:
1s
 - I think we should prioritise the elegance of the output over the elegance of the code in this instance since it's specifically a function which gives nicely formatted output!

Reply:
Ahhh, good point, my mistake there!